server/weakattr.py

```diff
@@ -0,0 +1,29 @@
+import contextlib
+import weakref
+from typing import Generic, Optional, TypeVar
+
+T = TypeVar("T")
+
+
+# Adapted from https://www.python.org/dev/peps/pep-0487/
+class WeakAttribute(Generic[T]):
+    """
+    Transparently allow an object attribute to reference another object via a
+    weak reference.
+    """
+
+    def __set_name__(self, _: object, name: str) -> None:
+        self.name = name
+
+    def __get__(self, obj: object, objclass: type) -> Optional[T]:
+        ref = obj.__dict__.get(self.name)
+        if ref:
+            return ref()
+        return None
```

Comment:
Here is a shorter alternative that might or might not be more readable :stuck_out_tongue_closed_eyes: 
```suggestion
        ref = obj.__dict__.get(self.name, lambda: None)
        return ref()
```

Reply:
I don't think it's unreadable enough to do that. Besides, my way is marginally faster because it avoids creating a temporary function and avoids one function call when the item isn't found :P 

```python
>>> timeit.timeit("""
... a = {}.get(1)
... if a:
...     a()
... """)
0.07964258699212223
>>> timeit.timeit("{}.get(1, lambda: None)()")
0.13527880399487913
>>> timeit.timeit("""
... a = {1: func}.get(1)
... if a:
...     a()
... """, "func = lambda: 10")
0.16154761402867734
>>> timeit.timeit("{1: func}.get(1, lambda: None)()", "func = lambda: 10")
0.18628269800683483
```

I added explicit `return None` because mypy was complaining, although it isn't technically necessary.

Reply:
touché